File: Thesis - Python/Genetic Algorithm/hga.py
from itertools import product,combinations
import numpy as np
from scipy.stats import qmc

class HGA:
    def __init__(
            self,
            systemeq,
            population_size,
            boundaries,
            max_generation,
            mutation_rate,
            parts,
            epsilon,
            delta,
            seed
            ) -> None:
        """
        PROPERTIES
            systemeq: func= systen of equation in question
            boundaries: np.ndarray= boundaries/constraint of the problem
            dim: int= dimension of roots
            population_size: int= population size per generation
            max_generation: int= maximal iteration
            max_memories_size: int=  maximum memories size
            mutation_rate: np.ndarray= probability of mutation
            epsilon: the error of the objective function
            delta: the minimum distance between roots
            seed: int= random state
        """
        self.population_size = population_size
        self.systemeq = systemeq
        self.boundaries = boundaries
        self.dim = boundaries.shape[0]
        self.max_generation = max_generation
        self.mutation_rate = mutation_rate
        self.parts = parts
        self.epsilon = epsilon
        self.delta = delta
        self.seed = seed
        return None
    
    def objective_function(self,x:np.ndarray):
        res = 0
        F_array = self.systemeq(x)
        for f in F_array:
            res += np.abs(f)
        self.condition = -1+self.epsilon
        return -1/(1+res)
    
        """GENERATE POINTS USING SOBOL SEQUENCE"""
    def generate_points(
            self,
            npoint: int, 
            low=-10, 
            high=10, 
            sobol=True
            ):
        """
                Generates points within the specified bounds.

            Args:
                dim: Number of dimensions.
                npoint: Number of points to generate.
                low: Lower bound for each variable (scalar or list/numpy array).
                high: Upper bound for each variable (scalar or list/numpy array).
                sobol: Flag indicating whether to use Sobol sequence (True) or random sampling (False).

            Returns:
                A numpy array of size (npoint, dim) representing the generated points.
        """

        if type(low) != type(high):
            raise TypeError('The type of "low" and "high" should be the same.')

        # Handle boundaries
        if type(low) == int:
            boundaries = [(low, high) for _ in range(self.dim)]
        elif type(low) in (list, np.ndarray):
            if len(low) != len(high):
                raise TypeError('The length of "low" and "high" should be the same.')
            else:
                boundaries = [(low[i], high[i]) for i in range(len(low))]

        # Generate points based on the sobol flag
        if sobol:
            sampler = qmc.Sobol(d=self.dim,scramble=True,seed=self.seed)
            sample = sampler.random(n=npoint)
            scaled_points = qmc.scale(sample=sample,l_bounds=low,u_bounds=high)

        else:
            # Generate random points
            np.random.seed(self.seed)
            scaled_points = np.zeros((npoint, self.dim))
            for i in range(self.dim):
                min_val, max_val = boundaries[i]
                scaled_points[:, i] = np.random.uniform(min_val, max_val, npoint)

        return scaled_points
    
    """Roullete wheel selection"""
    def selection(
            self,
            population: np.ndarray,
            fitness: np.ndarray):
        population_size = population.shape[0]
        selection_probs = np.array([1 / (fit + 1) for fit in fitness]) # add one to avoid negative probability
        total_probs = sum(selection_probs)
        selection_probs = np.array([prob / total_probs for prob in selection_probs])
        selected_indices = np.random.choice(a=np.arange(population_size),size=population_size,p=selection_probs)
        selected_population = np.array([population[i] for i in selected_indices])
        return selected_population
    
    """Single point crossover"""
    def crossover(self,parent1, parent2):
        dimension = len(parent1)
        crossover_point = np.random.randint(1, dimension)
        offspring1 = np.append(parent1[:crossover_point], parent2[crossover_point:])
        offspring2 = np.append(parent2[:crossover_point], parent1[crossover_point:])
        return [offspring1, offspring2]
    
    """One point mutation"""
    def mutate(self, individual, mutation_rate, boundaries):
        for j in range(len(individual)):
            if np.random.random() < mutation_rate:
                individual[j] = np.random.uniform(boundaries[j][0], boundaries[j][1])
        return individual
    
    def recombination(self,population: np.ndarray,mutation_rate, boundaries):
        offspring_population = []
        population_size = population.shape[0]
        for i in range(0, population_size, 2):
            parent1 = population[i]
            parent2 = population[i + 1]
            offspring1, offspring2 = self.crossover(parent1=parent1,parent2=parent2)
            offspring1 = self.mutate(individual=offspring1,mutation_rate=mutation_rate, boundaries=boundaries)
            offspring2 = self.mutate(individual=offspring2,mutation_rate=mutation_rate, boundaries=boundaries)
            offspring_population.extend([offspring1, offspring2])
        offspring_population = np.array(offspring_population)
        return offspring_population
    
    def GA(self,
           population_size,
           boundaries,
           max_generation,
           mutation_rate,
           seed=0,
           print_stat = False):
        
        np.random.seed(seed)
        dim = boundaries.shape[1]
        population = self.generate_points(npoint=population_size,low=boundaries[:,0],high=boundaries[:,1],sobol=True)
        fitness = np.asarray([self.objective_function(ind) for ind in population])
        best_idx = np.argmin(fitness)
        best_individual = population[best_idx]

        for generation in range(max_generation):
            selected_population = self.selection(population=population,fitness=fitness)
            offspring_population = self.recombination(population=selected_population,
                                                      mutation_rate=mutation_rate,
                                                      boundaries=boundaries)
            population = offspring_population
            fitness = np.asarray([self.objective_function(ind) for ind in population])
            best_idx = np.argmin(fitness)
            best_individual = population[best_idx]
            best_fitness = fitness[best_idx]
            if print_stat == True:

                print(f"=========Generation {generation}=========")
                print(f"Best Individual: {best_individual}")
                print(f"Best Score: {best_fitness}\n")

        return best_individual, best_fitness
    
    def slice_hypercube(self,lower_bounds, upper_bounds, interval):
        dim = len(lower_bounds)
        # Create a list of arrays, each containing points spaced h apart for each dimension
        points = [np.arange(lower_bounds[i], upper_bounds[i], interval[i]) for i in range(dim)]
        
        # Use meshgrid to create a grid of points in n-dimensional space
        grids = np.meshgrid(*points, indexing='ij')
        
        # Flatten and combine the grid points into a single 2D array
        grid_points = np.vstack([grid.ravel() for grid in grids]).T
        
        # Generate all vertices for smaller hypercubes within each grid cell
        offsets = np.array(list(product(*[[0, val] for val in interval])))
        res = np.array([grid_points + offset for offset in offsets])
        return res
    
    def clustering(self):
        # length of each parts in each dimension
        inc_int = (self.boundaries[:,1]-self.boundaries[:,0])/self.parts

        # Hypercubes
        hypercubes_edges = self.slice_hypercube(lower_bounds=self.boundaries[:,0],
                                                upper_bounds=self.boundaries[:,1],
                                                interval=inc_int)  
        
        cluster = []
        for hypercube_id in range(hypercubes_edges.shape[1]):
            X0 = hypercubes_edges[:,hypercube_id,:]
            F_list = self.systemeq(X0.T)

            # cek jika f yang berubah tanda dari F_list jika dievaluasi di tiap edge hypercube
            product_combination = np.array([[a*b for a,b in combinations(F_list[i],2)] for i in range (F_list.shape[0])])

            # jika semua f dari F_list berubah tanda jika dievaluasi di tiap edge hypercube, maka ada akar di situ
            change_sign = np.array([np.any(product_combination[i]<0) for i in range (product_combination.shape[0])])
            if np.all(change_sign==True):
                # print(f'Ada akar di sini: \nX0={X0}')
                cluster.append(X0)

        self.cluster = np.array(cluster)
        print(f"Number of clusters containing root: {self.cluster.shape[0]}")

    def root_elimination(self,root_archive):
        if self.dim == 1:
            list_criteria = [element for sublist in root_archive for element in sublist] #convert from 2D array into 1D array
        else:
            list_criteria = root_archive
        eligible_roots = np.array([x for x in list_criteria if (self.objective_function(x))<self.condition])
        id_duplicated_roots = []
        for i in range(len(eligible_roots)):
            for j in range (i+1,len(eligible_roots)):
                if np.linalg.norm(eligible_roots[i]-eligible_roots[j])<self.delta:
                    id_duplicated_roots.append([i,j])
        id_duplicated_roots = np.unique(id_duplicated_roots,axis=0)
        deselected_id_duplicated_roots = []
        for i in range (len(id_duplicated_roots)):
            root_a = self.objective_function(eligible_roots[id_duplicated_roots[i][0]])
            root_b = self.objective_function(eligible_roots[id_duplicated_roots[i][1]])
            if root_a<=root_b:
                id_duplicated_root = id_duplicated_roots[i][1]
            else:
                id_duplicated_root = id_duplicated_roots[i][0]
            deselected_id_duplicated_roots.append(id_duplicated_root)

        if deselected_id_duplicated_roots:
            unique_roots = np.ones(len(eligible_roots),dtype=bool)
            unique_roots[deselected_id_duplicated_roots] = False
            final_root = eligible_roots[unique_roots]
        else:
            final_root = eligible_roots
        return final_root

    def GA_evaluation(self,verbose = False, superverbose = False):
        self.archive = []
        self.score = []
        for i in range (self.cluster.shape[0]):
            subbound = np.array([[self.cluster[i,:,:][:,d].min(),self.cluster[i,:,:][:,d].max()] for d in range(self.cluster.shape[2])])
            root,root_score = self.GA(population_size=self.population_size,
                                 boundaries=subbound,
                                 max_generation=self.max_generation,
                                 mutation_rate=self.mutation_rate,
                                 seed = self.seed,
                                 print_stat=superverbose)
            self.archive.append(root)
            self.score.append(root_score)
            if verbose == True:
                print(f'\n====== Cluster {i+1} ======\n')
                print(f'Roots = {self.archive}')

        self.final_root = self.root_elimination(self.archive)

File: Thesis - Matlab/Hypercube/HGA.m
classdef HGA < handle
    properties
        boundaries
        dim
        population_size
        max_generation
        mutation_rate
        delta
        epsilon
        parts
        seed
        cluster
        archive
        score
        final_root
        final_score
    end
    methods
        function obj = HGA(boundaries, population_size, parts, max_generation,mutation_rate,epsilon, delta, seed)
            obj.boundaries = boundaries;
            obj.dim = size(boundaries,1);
            obj.population_size = population_size;
            obj.max_generation = max_generation;
            obj.parts = parts;
            obj.mutation_rate = mutation_rate;
            obj.epsilon = epsilon;
            obj.delta = delta;
            obj.seed = seed;
            obj.cluster = {};
            obj.archive = {};
            obj.score = [];
        end

        function F_array = system_equations(obj,x)
            f1 = exp(x(1)-x(2)) - sin(x(1)+x(2));
            f2 = (x(1)*x(2))^2 - cos(x(1)+x(2));
            F_array = [f1; f2];
        end

        function res = objective_function(obj, x)
            F_array = obj.system_equations(x);
            res = sum(abs(F_array));
            res = -1 / (1 + res);
        end
        
        function points = generate_points(obj,npoint,boundaries,seed)
            rng(seed)
            dimension = size(boundaries,1);
            p = sobolset(dimension);
            p = scramble(p,'MatousekAffineOwen');
            A = net(p,npoint);
            points = zeros(npoint,dimension);
            for i=1:dimension
               points(:,i)=round((boundaries(i,1)+(boundaries(i,2)-boundaries(i,1)).*A(:,i))*100)/100;
            end
        end

        function selected_population = selection(obj, population, fitness)
            pop_size = size(population, 1);
            selection_probs = 1 ./ (fitness + 1); % add one to avoid division by zero
            total_probs = sum(selection_probs);
            selection_probs = selection_probs / total_probs;
            selected_indices = randsample(1:pop_size, pop_size, true, selection_probs);
            selected_population = population(selected_indices, :);
        end

        function [offspring1, offspring2] = crossover(obj, parent1, parent2)
            dimension = length(parent1);
            crossover_point = randi([1, dimension], 1);
            offspring1 = [parent1(1:crossover_point), parent2(crossover_point+1:end)];
            offspring2 = [parent2(1:crossover_point), parent1(crossover_point+1:end)];
        end

        function individual = mutate(obj, individual, mutation_rate, boundaries)
            for j = 1:length(individual)
                if rand() < mutation_rate
                    individual(j) = rand() * (boundaries(j, 2) - boundaries(j, 1)) + boundaries(j, 1);
                end
            end
        end

        function offspring_population = recombination(obj, population, mutation_rate, boundaries)
            offspring_population = [];
            pop_size = size(population, 1);
            for i = 1:2:pop_size
                parent1 = population(i, :);
                parent2 = population(i + 1, :);
                [offspring1, offspring2] = obj.crossover(parent1, parent2);
                offspring1 = obj.mutate(offspring1, mutation_rate, boundaries);
                offspring2 = obj.mutate(offspring2, mutation_rate, boundaries);
                offspring_population = [offspring_population; offspring1; offspring2];
            end
        end

        function [best_individual, best_fitness] = GA(obj, population_size, boundaries, max_generation, mutation_rate, seed, print_stat)
            rng(seed);
            dimension = size(boundaries, 2);
            population = obj.generate_points(population_size, boundaries, seed);
            fitness = zeros(1, population_size);
            for i = 1:population_size
                fitness(i) = obj.objective_function(population(i, :));
            end
            [best_fitness, best_idx] = min(fitness);
            best_individual = population(best_idx, :);
        
            for generation = 1:max_generation
                selected_population = obj.selection(population, fitness);
                offspring_population = obj.recombination(selected_population, mutation_rate, boundaries);
                population = offspring_population;
                fitness = zeros(1, population_size);
                for i = 1:population_size
                    fitness(i) = obj.objective_function(population(i, :));
                end
                [best_fitness, best_idx] = min(fitness);
                best_individual = population(best_idx, :);
                if print_stat == true
                    disp("=========Generation " + generation + "=========");
                    disp("Best Individual: ");
                    disp(best_individual);
                    disp("Best Score: ");
                    disp(best_fitness);
                    disp(" ");
                end
            end
        end

        function res = slice_hypercube(obj, lower_bounds, upper_bounds, parts)
            interval = (upper_bounds-lower_bounds)/(parts);
            dimension = numel(lower_bounds);
            
            % Create a cell array of arrays, each containing points spaced interval apart for each dimension
            points = cell(1, dimension);
            for i = 1:dimension
                points{i} = linspace(lower_bounds(i), upper_bounds(i)-interval(i), parts);
            end
            
            % Use ndgrid to create a grid of points in n-dimensional space
            [grids{1:dimension}] = ndgrid(points{:});
            
            % Combine the grid points into a single matrix
            grid_points = reshape(cat(dimension+1, grids{:}), [], dimension);
            
            % % Generate all vertices for smaller hypercubes within each grid cell
            zero_interval = cell(dimension,1);
            for i = 1:dimension
                zero_interval{i} = [0,interval(i)];
            end
            
            [offsets_grid{1:dimension}] = ndgrid(zero_interval{:});
            offsets_grid_points = reshape(cat(dimension+1, offsets_grid{:}), [], dimension);
            res = zeros([size(offsets_grid_points, 1),size(grid_points)]);
            for i = 1:size(offsets_grid_points, 1)
                res(i,:,:) = grid_points + offsets_grid_points(i,:);
            end
        end

        function clustering(obj)
            lower_bounds = obj.boundaries(:,1); upper_bounds = obj.boundaries(:,2); 
            interval = (upper_bounds-lower_bounds)/(obj.parts);
            
            hypercubes_edges = obj.slice_hypercube(lower_bounds,upper_bounds,obj.parts);
            
            for hypercube_id = 1:size(hypercubes_edges, 2)
                X0 = hypercubes_edges(:,hypercube_id,:);
                X0 = reshape(X0, [], size(X0, 3));
                
                F_list = zeros(obj.dim,size(X0,1));
                for i = 1:size(X0,1)
                    F_list(:,i) = obj.system_equations(X0(i,:));
                end
                
                product_combination = zeros(size(F_list, 1), nchoosek(size(F_list, 2), 2));
                for i = 1:size(F_list, 1)
                    combinations = nchoosek(F_list(i,:), 2);
                    product_combination(i,:) = prod(combinations, 2)';
                end
                
                change_sign = any(product_combination < 0, 2);
                if all(change_sign)
                    obj.cluster{end+1} = X0;
                end
            end
            % Convert cluster to array
            obj.cluster = cat(3, obj.cluster{:});
        end

        function final_root = root_elimination(obj,root_archive)
            eligible_roots = [];
            for i = 1:size(root_archive,1)
                if obj.objective_function(root_archive(i,:)) < -1 + obj.epsilon
                    eligible_roots = [eligible_roots; root_archive(i,:)];
                end
            end
            
            id_duplicated_roots = [];
            for i = 1:length(eligible_roots)
                for j = i+1:length(eligible_roots)
                    if norm(eligible_roots(i,:) - eligible_roots(j,:)) < obj.delta
                        id_duplicated_roots = [id_duplicated_roots; [i, j]];
                    end
                end
            end
            
            id_duplicated_roots = unique(id_duplicated_roots, 'rows');
            
            deselected_id_duplicated_roots = [];
            for i = 1:size(id_duplicated_roots, 1)
                root_a = obj.objective_function(eligible_roots(id_duplicated_roots(i, 1),:));
                root_b = obj.objective_function(eligible_roots(id_duplicated_roots(i, 2),:));
                if root_a <= root_b
                    id_duplicated_root = id_duplicated_roots(i, 2);
                else
                    id_duplicated_root = id_duplicated_roots(i, 1);
                end
                deselected_id_duplicated_roots = [deselected_id_duplicated_roots; id_duplicated_root];
            end
            
            if ~isempty(deselected_id_duplicated_roots)
                unique_roots = true(size(eligible_roots,1),1);
                unique_roots(deselected_id_duplicated_roots) = false;
                final_root = eligible_roots(unique_roots,:);
            else
                final_root = eligible_roots;
            end
        end
        
        function [final_root,final_score] = GA_evaluation(obj, verbose, superverbose)
            obj.clustering()
            if verbose == true
                fprintf("Number of clusters containing root: %d\n", size(obj.cluster, 3));
            end
            for i = 1:size(obj.cluster, 3)
                subbound = zeros(obj.dim, 2);
                for d = 1:size(obj.cluster, 2)
                    subbound(d, :) = [min(obj.cluster(:, d, i)), max(obj.cluster(:, d, i))];
                end
        
                [root, root_score] = obj.GA(obj.population_size,subbound,obj.max_generation,obj.mutation_rate,obj.seed,superverbose);
                
                obj.archive{end+1} = root;
                obj.score(end+1) = root_score;
                
                if verbose
                    fprintf('\n====== Cluster %d ======\n', i);
                    disp('Roots =');
                    disp(obj.archive);
                end
            end

            % Convert cell array to matrix
            archive_matrix = zeros(numel(obj.archive), numel(obj.archive{1}));
            for i = 1:numel(obj.archive)
                archive_matrix(i, :) = obj.archive{i};
            end

            final_root = root_elimination(obj, archive_matrix);
            final_score = zeros(1, size(final_root,1));
            for fin_iter = 1:size(final_root,1)
                final_score(fin_iter) = obj.objective_function(final_root(fin_iter, :));
            end
            obj.final_root = final_root;
            obj.final_score = final_score;
        end
        
        function visualization2D(obj,visual_properties)
            if visual_properties.save_visual == true
                writerObj = VideoWriter(visual_properties.file_name);
                writerObj.FrameRate = 5;  % Adjust the frame rate as needed
                open(writerObj);
            end

            % Create a figure with visibility off
            if visual_properties.show_visual == false
                fig = figure('Visible', 'off');
            else 
                fig = figure('Visible', 'on');
            end

            % Adjust aspect ratio
            axis equal;
            pbaspect([diff(xlim()) diff(ylim()) 1]);
            
            % Maximize figure window
            set(gcf, 'WindowState', 'maximized');

            xlim(obj.boundaries(1,:));
            ylim(obj.boundaries(2,:));
            hold on
            rectangle('Position',[obj.boundaries(:,1)',(obj.boundaries(:,2)-obj.boundaries(:,1))'],'EdgeColor','#FF0000')
            for i = 1:size(obj.cluster, 3)
                subbound = zeros(obj.dim, 2);
                for d = 1:size(obj.cluster, 2)
                    subbound(d, :) = [min(obj.cluster(:, d, i)), max(obj.cluster(:, d, i))];
                end
                
                rectangle('Position',[subbound(:,1)',(subbound(:,2)-subbound(:,1))'],'EdgeColor','#4DBEEE')
                pause(0.25)
                if visual_properties.save_visual == true
                    frame = getframe(gcf);
                    writeVideo(writerObj, frame);
                end
            end

            for j = 1: size(obj.final_root,1)
                plot(obj.final_root(j,1), obj.final_root(j,2), '*',Color='magenta',MarkerSize=50);
                pause(0.25)
                if visual_properties.save_visual == true
                    frame = getframe(gcf);
                    writeVideo(writerObj, frame);
                end
            end
            hold off

            if visual_properties.save_visual == true
                close(writerObj);
            end

        end


    end
end